lib/diagnostics_channel.js

```diff
@@ -0,0 +1,141 @@
+'use strict';
+
+const {
+  ObjectDefineProperty,
+  ObjectSetPrototypeOf,
+  ObjectValues,
+} = primordials;
+
+const {
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+  }
+} = require('internal/errors');
+
+// TODO(qard): should there be a C++ publisher interface?
+class ActivePublisher {
+  subscribe(subscription) {
+    if (typeof subscription === 'function') {
+      this._subscribers.push(subscription);
+      return;
+    }
+
+    if (typeof subscription === 'object' &&
+        typeof subscription.onMessage === 'function') {
+      let shouldListen = true;
+
+      if (typeof subscription.shouldListen === 'function') {
+        shouldListen = subscription.shouldListen(this.name);
+      }
+
+      if (shouldListen) {
+        this._subscribers.push(subscription.onMessage.bind(subscription));
+      }
+    }
+  }
+
+  shouldPublish() {
+    return true;
+  }
+
+  publish(data) {
+    for (var i = 0; i < this._subscribers.length; i++) {
+      this._subscribers[i](this.name, data);
+    }
+  }
+}
+
+class Publisher {
+  constructor(name) {
+    this._subscribers = undefined;
+    ObjectDefineProperty(this, 'name', {
+      value: name
+    });
+  }
+
+  subscribe(subscription) {
+    ObjectSetPrototypeOf(this, ActivePublisher.prototype);
+    this._subscribers = [];
+    this.subscribe(subscription);
+    ObjectDefineProperty(this, 'ifActive', {
+      value: this
```

Comment:
default descriptor is `configurable: false, writable: false`, this seems off

Reply:
Nope, that's intentional. It transitions to active the first time something subscribes to it. When that happens, it swaps the prototypes and sets `ifActive` to no longer be `undefined`, that way you can then do `publisher.ifActive?.publish(...)` and it will succeed. Before that point, it would skip because of being undefined until then. Perhaps not the clearest code though. 🤔

Reply:
I switched this to be a plain property too, and included an undefined initialization in the constructor to make it more clear that this point is a state transition.